### PR TITLE
CONTRIBUTING: rebase master to staging without pinging people

### PR DESCRIPTION
**Pull request by Mic92**
###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by dasJ** [comment]
cc @jonringer since this is based on https://github.com/NixOS/nixpkgs/pull/144227#pullrequestreview-800648130 ;)

**Review by dasJ — APPROVED** [review]

**Review by 0x4A6F — APPROVED** [review]

**Review by Artturin — APPROVED** [review]
i've been using this made by jonringer
```
# finding the common merge base will avoid pinging all CODEOWNERs
common=$(git merge-base origin/master origin/staging)
git rebase --onto=$common HEAD~1 # or however many commits you want to pull
git push --force
```
now i wont have to dig up my notes everytime hehe

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -52,6 +52,35 @@ In addition to writing properly formatted commit messages, it's important to inc
...
+commits=$(git merge-base $(git branch --show-current) upstream/master)
+# Rebase all commits onto the common base
+git rebase --onto=$common $commits
+# Force push your changes
+git push origin $(git branch --show-current) --force-with-lease
+```
+
+Then change the base branch in the GitHub PR with the "Edit" button in the top
+right corner, i.e. `master` → `staging`.
+
```
Comment (piegamesde) [inline CONTRIBUTING.md:83]: Might be worth adding that one should rebase onto latest staging afterwards.

Comment (mweinelt) [reply CONTRIBUTING.md:83]: Indeed there might be lingering changes between the merge base and the latest changes on staging that may require that rebase.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -52,6 +52,35 @@ In addition to writing properly formatted commit messages, it's important to inc
...
+
+```console
+# Find the common base between two branches
+common=$(git merge-base upstream/master upstream/staging)
+# Find the common base between your feature branch and master
+commits=$(git merge-base $(git branch --show-current) upstream/master)
+# Rebase all commits onto the common base
+git rebase --onto=$common $commits
+# Force push your changes
```
Comment (dasJ) [inline CONTRIBUTING.md:77]:
```suggestion
# Force push your changes
git fetch upstream staging
git rebase upstream/staging
```

Comment (jonringer) [reply CONTRIBUTING.md:77]: this shouldn't be necessary since they should have done `git rebase --onto=$common $commits`

**Review by jonringer — APPROVED** [review]

File: CONTRIBUTING.md (hunk 2 above)
Comment (piegamesde) [reply CONTRIBUTING.md:77]: Rebasing onto upstream/staging should only be done after having force-pushed *and* switched the target branch in GitHub (see https://github.com/NixOS/nixpkgs/pull/166472#discussion_r838982527). (I may be misunderstanding your suggestion though)

Comment (mweinelt) [reply CONTRIBUTING.md:77]: The merge base between master and staging is usually older than staging itself, and in between that diff there may be conflicting changes still.

Comment (jonringer) [reply CONTRIBUTING.md:77]: oh yea, that's fair. Keep forgetting about divergence :(

File: CONTRIBUTING.md (hunk 1 above)
Comment (SuperSandro2000) [reply CONTRIBUTING.md:83]: The very important part here is to do it after changing the base branch.

**Comment by K900** [comment]
Feels like this could be solved in ofborg instead, at least for the most common cases? Just have a list of valid branch targets (master, staging, release-XX.YY), then pick the one with lowest distance to merge base for the pings.

**Comment by dasJ** [comment]
@K900 that's very much out of scope for this PR

**Comment by mweinelt** [comment]
Updated to reflect reviewer feedback. PTAL.

**Review by piegamesde — APPROVED** [review]

**Comment by dasJ** [comment]
cc @siraben because of #166605

**Comment by Artturin** [comment]
The fixup commit wasn't fixed up

**Comment by Artturin** [comment]
https://github.com/NixOS/nixpkgs/pull/167312

**Comment by ncfavier** [comment]
Simplified the instructions in https://github.com/NixOS/nixpkgs/pull/175352.

**Pull request merged**